misk/src/main/kotlin/misk/digester/MergingDigest.kt

```diff
@@ -0,0 +1,351 @@
+      )
+      if (swappedCentroids.size != 0) {
+        nextMain = swappedCentroids[0]
+      } else if (actualMainCentroids.size != 0) {
+        nextMain = actualMainCentroids[0]
+      }
+
+      if (nextMain.mean < nextTemp.mean) {
+        if (actualMainCentroids.size != 0) {
+          if (swappedCentroids.size != 0) {
+            // if this came from swap, before merging, we have to save
+            // the next main centroid at the end
+            // this copy is probably the most expensive part of the
+            // in-place merge, compared to merging into a separate buffer
+            swappedCentroids.removeAt(0)
+            swappedCentroids[(swappedCentroids.size)-1] = actualMainCentroids[0]
+          }
+          actualMainCentroids.removeAt(0)
+        } else {
+          // the real main has been completely exhausted, so we're just
+          // cleaning out swapped mains now
+          swappedCentroids.removeAt(0)
+        }
+
+        lastMergedIndex = mergeOne(mergedWeight, totalWeight, lastMergedIndex, nextMain)
+        mergedWeight += nextMain.weight
+      } else {
+        // before merging, we have to save the next main centroid somewhere
+        // else, so that we don't overwrite it
+        if (actualMainCentroids.size != 0) {
+          swappedCentroids.add(actualMainCentroids[0])
+          actualMainCentroids.removeAt(0)
+        }
+        tempIndex++
+
+        lastMergedIndex = mergeOne(mergedWeight, totalWeight, lastMergedIndex, nextTemp)
+        mergedWeight += nextTemp.weight
+      }
+    }
+
+    tempCentroids.clear()
+    tempWeight = 0.0
+    mainWeight = totalWeight
+  }
+
+  /**
+   * merges a single centroid into the mergedCentroids list
+   * note that "merging" sometimes creates a new centroid in the list, however
+   * the length of the list has a strict upper bound (see constructor)
+   */
+  fun mergeOne(
+      beforeWeight: Double,
+      totalWeight: Double,
+      beforeIndex: Double,
+      next: Centroid
+  ): Double {
+
+    // compute the quantile index of the element we're about to merge
+    var nextIndex = indexEstimate((beforeWeight + next.weight) / totalWeight)
+
+    if (nextIndex-beforeIndex > 1 || mainCentroids.size == 0) {
+      // the new index is far away from the last index of the current centroid
+      // therefore we cannot merge into the current centroid or it would
+      // become too wide, so we will append a new centroid
+      mainCentroids.add(next)
+      // return the last index that was merged into the previous centroid
+      return indexEstimate(beforeWeight / totalWeight)
+    } else {
+      // the new index fits into the range of the current centroid, so we
+      // combine it into the current centroid's values
+      // this computation is known as welford's method, the order matters
+      // weight must be updated before mean
+
+      //TODO: is this thread safe? doublecheck that this successfully updates all properties
```

Comment:
This class isn't thread-safe in general. That's ok, though. We'll just need to make sure that whatever uses it deals with concurrency.